### PR TITLE
Federated Dataset Tutorial 

### PR DESCRIPTION
**Pull request by kamathhrishi**
Addresses issue #2036 

A tutorial which makes it easy to create datasets that work with federated data loaders. Our previous tutorials have used torchvision datasets or small toy datasets which were compatible with federated data loaders. This notebook still requires some refining. 
@LaRiffle 

Let me know of any additions/changes I could make. Anything that makes it easier to use BaseDataset and federated data loaders.

### CODE REVIEW

File: syft/workers/websocket_client.py [hunk 1]
```diff
@@ -16,7 +16,7 @@
 logger = logging.getLogger(__name__)
 
 
-TIMEOUT_INTERVAL = 9999999
+TIMEOUT_INTERVAL = 9_999_999
```
Comment (mari-linhares) [inline syft/workers/websocket_client.py]: Was this change intended?

File: examples/tutorials/advanced/Federated Dataset.ipynb [hunk 2]
```diff
@@ -0,0 +1,460 @@
...
+   "metadata": {},
+   "source": [
+    "<h1>Federated Datasets</h1>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This example demonstrates how you can develop your custom datasets that work with federated data loaders. The regular torch datasets do not always work directly with federated data loaders.The tutorial aims to highlight the differences and shows you how you could develop a federated dataset with your data that works with the federated data loader. Further you could also use Syft's Base dataset feature which simplify creating federated datasets.To demonstrate these we will show you how to load the SVHN (Street View House Numbers) dataset and convert it to a federated dataset.\n",
```
Comment (mari-linhares) [inline examples/tutorials/advanced/Federated Dataset.ipynb]: There is "federated data loaders" many times in different sentences maybe...

Instead of:
> This example demonstrates how you can develop your custom datasets that work with federated data loaders. The regular torch datasets do not always work directly with federated data loaders.The tutorial aims to highlight the differences and shows you how you could develop a federated dataset with your data that works with the federated data loader. Further you could also use Syft's Base dataset feature which simplify creating federated datasets.

I would write something like:
*Unfortunately the regular torch datasets do not always work directly with federated data loaders. This example demonstrates how to develop a custom dataset that works with federated data loaders highlighting the differences. Further, you could also use Syft's Base dataset feature which simplifies creating federated datasets.*

Also, can you add a link to the official SVHN dataset?

File: examples/tutorials/advanced/Federated Dataset.ipynb [hunk 3]
```diff
@@ -0,0 +1,460 @@
...
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "class SVHNDataset(Dataset):\n",
+    "\n",
+    "    def __init__(self,images,labels,transform=None):\n",
+    "        \n",
+    "        \"\"\"arguments:\n",
```
Comment (mari-linhares) [inline examples/tutorials/advanced/Federated Dataset.ipynb]: arguments -> Args

File: examples/tutorials/advanced/Federated Dataset.ipynb [hunk 4]
```diff
@@ -0,0 +1,460 @@
...
+    "             \n",
+    "             Images (Numpy Array): Text corpus \n",
```
Comment (mari-linhares) [inline examples/tutorials/advanced/Federated Dataset.ipynb]: Images -> images. All params should be written as in the original arguments.

Comment (mari-linhares) [reply examples/tutorials/advanced/Federated Dataset.ipynb]: Also, add a description for every argument and finish a sentence with a period.

File: examples/tutorials/advanced/Federated Dataset.ipynb [hunk 5]
```diff
@@ -0,0 +1,460 @@
...
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "class Arguments():\n",
+    "    def __init__(self):\n",
+    "        self.batch_size = 1\n",
```
Comment (LaRiffle) [inline examples/tutorials/advanced/Federated Dataset.ipynb]: You don't need all these like the .lr I think

File: examples/tutorials/advanced/Federated Dataset.ipynb [hunk 6]
```diff
@@ -0,0 +1,460 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 13,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "federated_train_loader = sy.FederatedDataLoader( # <-- this is now a FederatedDataLoader \n",
```
Comment (LaRiffle) [inline examples/tutorials/advanced/Federated Dataset.ipynb]: Maybe distinguish the two phases: 1) .federate and 2) put into a federatedDataLoader

File: examples/tutorials/advanced/Federated Dataset.ipynb [hunk 7]
```diff
@@ -0,0 +1,460 @@
...
+       "<syft.frameworks.torch.federated.dataset.BaseDataset at 0x18196b4828>"
+      ]
+     },
+     "execution_count": 14,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "base.fix_prec()"
```
Comment (LaRiffle) [inline examples/tutorials/advanced/Federated Dataset.ipynb]: That's sompething that you maybe want to send more in depth for people to understand what's happening

File: syft/workers/websocket_client.py (hunk 1 above)
Comment (kamathhrishi) [reply syft/workers/websocket_client.py]: No , pretty surprising how it ended up.

File: examples/tutorials/advanced/Federated Dataset.ipynb (hunk 2 above)
Comment (kamathhrishi) [reply examples/tutorials/advanced/Federated Dataset.ipynb]: Thanks that simplified it :)

File: examples/tutorials/advanced/Federated Dataset.ipynb (hunk 5 above)
Comment (kamathhrishi) [reply examples/tutorials/advanced/Federated Dataset.ipynb]: yeah, I guess its unnecessary.

File: syft/workers/websocket_client.py (hunk 1 above)
Comment (mari-linhares) [reply syft/workers/websocket_client.py]: I think it was `black`.

Comment (kamathhrishi) [reply syft/workers/websocket_client.py]: Yep , I guess let it remain that way.

File: examples/tutorials/advanced/Federated Dataset.ipynb [hunk 8]
```diff
@@ -0,0 +1,473 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "args=Arguments()"
```
Comment (mari-linhares) [inline examples/tutorials/advanced/Federated Dataset.ipynb]: fix style -> args = Arguments()

File: examples/tutorials/advanced/Federated Dataset.ipynb [hunk 9]
```diff
@@ -0,0 +1,473 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "train_data,train_labels=load_data(\"data/train_32x32.mat\")\n",
```
Comment (mari-linhares) [inline examples/tutorials/advanced/Federated Dataset.ipynb]: fix style: add spaces.

File: examples/tutorials/advanced/Federated Dataset.ipynb [hunk 10]
```diff
@@ -0,0 +1,473 @@
...
+    "test_data,test_labels=load_data(\"data/test_32x32.mat\")"
```
Comment (mari-linhares) [inline examples/tutorials/advanced/Federated Dataset.ipynb]: fix style: add spaces.

File: examples/tutorials/advanced/Federated Dataset.ipynb [hunk 11]
```diff
@@ -0,0 +1,473 @@
...
...8BdGoU6R7TWKT2uSsoXIm9vqZZsRJR+Ftr5yynFtM9tNwu6jNJY0k0iKmf2H8x\ndSBzmnMM6Pl+LyVJPU70PUc/WZu0Dybg83pabtF3yupGuj/K3Xleu7bD8dFx5sulVs9761SlJz5a\nkShCaWFcT1rb0CqYCCOSPGn5S2k8Ko1FfIyVIniJvvAof3VDF1Qr6dNZ6tiHrue3dEp7debTNBgM\nBoPhemAfTYPBYDAYNsSWzbMeTbtCCR2+HWtXAMi1bk9GF6ksrqunR/MJeUmVqWAuIdRlGZYPHtyL\n++4dBHagxYmEymeBAcFhrRkrzmKYNG0/KqjEsfIFBm3Xd9HfF42Mfmi63TV0nc9MHymAYsjAwnUX\n83DSeWJ1BfSOgTbrDqOqGOMVq48oW06z5rhR1ylYpSQcN6mHgUAMFtMBZCmFiDatdM5+aktuFtod\nUx4QQvKnk1nGtMKAKvbsWPDUWJ+5Xp9lVtYLTWk5axOy38kyK3LMgMGWJ4j7WhkDrZgGucy3yTHq\nXdANZDkM8tsFdF2H09NF9k5jAFSqxJSO7wdVjb2bUkWT1M/RNbEO59b8ywwEauR4HdjDVJ68cLu4\naDBsQwwQZPs0J3DuDcjE1C/MMlYJZ1OYpmkwGAwGw4bYbpUT79GsWnSF1gbE+S4zVq9mLwxpZoCN\nno3yOG4r2nTOSSnBITLpnanAo4msN700hnCukVpwLXkMeQ5dtYPnOJ9fNNbvHJmojmlOu4YBuUHk\nhx1qmrHCSIz+1gEBXOEuFQYv52Jv5wFEMuOUnVngggSn1LVWNXOzgE4loAZVxTD4YYWWvpVAr4+T\nKVwuyOCm4ZxDVVUo9D30ZuIb91nMTUC+xMV9lghBJJBLp5dwfKnnnalo1BwVjWm0NjAQSP+OKpYf\n4SrtS+1FNJObROc7nJ2dZf3LYB0+A/kQveCd1Au00u/JflDVUnPPrlbZ8Zrbm1zCed3hXrUSbfGJ\n6xIcqF6skXyB2qeujnKxOfBSME3TYDAYDIYNsVVNsyhKHB4eZhUNYl3MOoT7kwIPSLOXZh1s4qyd\nCSQf2L2DQI6gU0GePHkCQIWSLxLFXrEfUk44Q16reoqk4pvtpaooR0+Fku9lmUE1ql7jmcxeRXXK\nasLFKuJh2XZaG+UMb7e0kDE4l89JoybWcTni0yzotzy/
```
Comment (mari-linhares) [inline examples/tutorials/advanced/Federated Dataset.ipynb]: fix style: add spaces.

Comment (mari-linhares) [inline examples/tutorials/advanced/Federated Dataset.ipynb]: fix style -> columns * rows + 1

Comment (mari-linhares) [inline examples/tutorials/advanced/Federated Dataset.ipynb]: fix style: add spaces.

Comment (mari-linhares) [inline examples/tutorials/advanced/Federated Dataset.ipynb]: fix style: add spaces.

Comment (mari-linhares) [inline examples/tutorials/advanced/Federated Dataset.ipynb]: fix style: add spaces.

Comment (mari-linhares) [inline examples/tutorials/advanced/Federated Dataset.ipynb]: fix style: add spaces.

Comment (mari-linhares) [inline examples/tutorials/advanced/Federated Dataset.ipynb]: fix style: add spaces.

Comment (mari-linhares) [inline examples/tutorials/advanced/Federated Dataset.ipynb]: fix style: add spaces.

Comment (mari-linhares) [inline examples/tutorials/advanced/Federated Dataset.ipynb]: fix style: add spaces.

Comment (mari-linhares) [inline examples/tutorials/advanced/Federated Dataset.ipynb]: fix style: add spaces.

Comment (mari-linhares) [inline examples/tutorials/advanced/Federated Dataset.ipynb]: fix style: add spaces.

Comment (mari-linhares) [inline examples/tutorials/advanced/Federated Dataset.ipynb]: fix style: add spaces.

Comment (mari-linhares) [inline examples/tutorials/advanced/Federated Dataset.ipynb]: fix style: add spaces args.

Comment (kamathhrishi) [reply examples/tutorials/advanced/Federated Dataset.ipynb]: Do these come under visualize dataset?

File: syft/workers/websocket_client.py (hunk 1 above)
Comment (robert-wagner) [reply syft/workers/websocket_client.py]: Apparently this is a valid pythonic way of writing an int (pep 515). Learned something new today

Comment (kamathhrishi) [reply syft/workers/websocket_client.py]: Interesting :)
